sw/device/mask_rom/boot.md

```diff
@@ -1,71 +1,201 @@
-# Pseudo-code for Secure Boot Process.
+# Pseudo-code for Mask ROM Secure Boot Process
 
-This file should be read in conjunction with secure boot specification. Some
-references to that document are included.
+This file should be read in conjunction with secure boot specification.
+References to that document are included.
 
+Sub-procedures are documented below the main `boot` function.
 
-1. Power on (entirely in hardware) Open Q: Whether SW has to configure PMP
-  initial region.
+1. Power on (entirely in hardware)
 
-2. Execution Begins
-  - CRT Startup Code (ASM)
-    - Clean Slate (SRAM, Registers)
-    - Well-defined interrupt/exception handler
-        Open Q: Do we want to enable interrupts in mask rom?
-    - Jump into C code (`boot`):
+   **Open Q:** Whether SW has to configure PMP initial region.
+
+2. Execution Begins with Mask ROM stage:
+   *   CRT Startup Code (Written in Assembly)
+
+       *   Disable Interrupts and set well-defined exception handler.
+           This should keep initial execution deterministic.
+
+       *   Clean Device State Part 1. (See "Cleaning Device State" below)
+           This includes enabling SRAM Scrambling.
+
+       *   Setup structures for C execution (CRT: `.data`, `.bss` sections, stack).
+
+       *   Jump into C code
+
+   *   Main Mask ROM Secure Boot Software (Written in C)
+       *   Orchestrated from `boot`:
 
 ```c
 void boot(void) {
-  // - Clean all other hardware state (using DIF Reset functions)
-  clean_device_state();
-
-  // - Chip-specific startup functionality
-  //   - Clocks
-  //   - AST
-  //   - Flash
-  //   - Entropy
-  //   Open Q: Proprietary Software Strategy.
+  boot_reason = read_boot_reason();
+
+  // Clean Device State Part 2.
+  // See "Cleaning Device State" Below.
+  clean_device_state_part_2(boot_reason);
+
+  // Enable Memory Protection
+  // - PMP Initial Region (if not done in power on)
+  enable_memory_protection();
+
+  // Chip-specific startup functionality
+  // **Open Q:** Proprietary Software Strategy.
+  // - Clocks
+  // - AST / Entropy. **Open Q:** Is this needed for SRAM Scrambling?
+  // - Flash
+  // - Fuses
   chip_specific_startup();
 
-  // - Read Boot Policy for ROM_EXT from Flash (2.b)
-  //   Open Q: How is this protected beyond being in flash?
-  boot_policy = read_boot_policy();
+  // Manufacturing boot-strapping intervention.
+  // **Open Q:** How does the mask rom deal with chip recovery?
+  // **Open Q:** Pin Strapping Configuration.
+  manufacturing_boot_strap();
 
-  // - Determine which rom_ext slot is in use (2.b,2.c.i)
-  rom_ext_manifest = read_rom_ext_slot(boot_policy);
+  // Read Boot Policy for ROM_EXT from Flash (2.b)
+  // **Open Q:** How is this protected beyond being in flash?
```

Comment:
i feel like the boot policy itself should part of the verification of the flash payload...is that not the case?

Reply:
This is the boot policy that you read to find out which ROM_EXT to boot, and what to do if you cannot boot that ROM_EXT. It has to be stored in flash (because it needs to be retained long-term), but I don't think it's stored in the main flash pages that the rom_ext manifests and images are stored in.